### ISSUE TITLE
Request to add stable sorting algorithm to PyTorch

### ISSUE THREAD

**Issue by seva100** [pytorch/pytorch#28871]
## 🚀 Feature
My feature request is to add the algorithm of [stable sorting](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability) to PyTorch. Stable sorting is a method of sorting an array that preserves order of elements with equal values. Formally, if there is a 1D tensor `A` and for some `i, j` `A[i] == A[j]`, then I would like to have `A[i]` earlier in the tensor than `A[j]` after its sorting. Generally speaking, the current implementation of `torch.sort` seems to be not preserving this order; for instance, see (produced by PyTorch 1.3.0):

```
In [2]: x = torch.Tensor([1, 2, 5, 4, 5])

In [3]: x.sort()
Out[3]: 
torch.return_types.sort(
values=tensor([1., 2., 4., 5., 5.]),
indices=tensor([0, 1, 3, 2, 4]))

In [4]: x.cuda().sort()
Out[4]: 
torch.return_types.sort(
values=tensor([1., 2., 4., 5., 5.], device='cuda:0'),
indices=tensor([0, 1, 3, 4, 2], device='cuda:0'))
```

## Motivation

This feature request is related to a problem I am facing outlined in https://discuss.pytorch.org/t/sorting-2d-tensor-by-pairs-not-columnwise/59465/8. Basically, I need to sort a 2D tensor of shape `(N, 2)` considering rows as tuples / pairs (row `A[i]` is less than row `A[j]`, iff `(A[i, 0] < A[j, 0]) or ((A[i, 0] == A[j, 0]) and (A[i, 1] <= A[j, 1]))`. There is a natural way to do this by sorting `A` by second column then stable sorting by first column. Please see https://discuss.pytorch.org/t/sorting-2d-tensor-by-pairs-not-columnwise/59465/8 for more detail.

If stable sort is implemented, it can be naturally extended to the sorting of tuples of any dimension via [Radix Sort](https://en.wikipedia.org/wiki/Radix_sort).

This issue seems to be related to a request about adding a stable sort to `torch.topk`: https://github.com/pytorch/pytorch/issues/27542 

## Pitch

Ideally, I would like to have a keyword argument in `torch.sort` or `torch.argsort`, for example `torch.sort(A, stable=True)`. There is already a similar syntax in NumPy: `np.sort(A, kind='stable')`. Stable sorting likely requires a different sorting algorithm than the one implemented in `torch.sort`.

## Alternatives

There is a simpler way to solve my problem suggested by @albanD but it does not suit my needs because of the possible precision loss (see https://discuss.pytorch.org/t/sorting-2d-tensor-by-pairs-not-columnwise/59465/8). If there is any other efficient solution to my problem that does not require stable sorting, please kindly post it here.

## Additional context

Here https://discuss.pytorch.org/t/is-torch-sort-stable/20714/3 is a small discussion about whether `torch.sort` is stable or not, and there is an answer 

> if the tensor contains at least 2049 elements, internally the function sortViaThrust is used, which as can be seen in the same file performs a stable sort.

I hope this can be useful.

cc @mruberry @rgommers

**Comment by colesbury** [comment]
@ngimel says that Thrust has a stable sort version of sorting. 

@dreiss says Java has "Tim" sort, which is stable, which might be a useful reference.

**Comment by pandeykartikey** [comment]
Hi, If this feature is still required. I would like to work on it.

**Comment by Balandat** [comment]
This would be great to have.

**Comment by andrew-bydlon** [comment]
Is there any progress on this?  This would also solve https://github.com/pytorch/pytorch/issues/3025 very naturally.

**Comment by mruberry** [comment]
No progress that I'm aware of, sorry.

**Comment by Karbo123** [comment]
I write a simple torch library that implements stable sorting. But it's still under development.
It is available at: https://github.com/Karbo123/torch_indexing
Welcome to try.

**Comment by ngimel** [comment]
Stable sort is implemented on cpu https://pytorch.org/docs/master/generated/torch.sort.html?highlight=sort#torch.sort, we'll work on enabling it on gpu too.

**Comment by ngimel** [comment]
Stable sort on GPU is now enabled.

**Comment by ZaydH** [comment]
> Stable sort on GPU is now enabled.

@ngimel The docs you cite only still says that stable only applies to CPU not GPU.  I assume this is out of date in the docs and needs to be updated.  Wanted to confirm before I issued the PR.

**Comment by ngimel** [comment]
Docs PR is merged, probably docs didn't update yet #61685
